README.md
# Match Video

This is a Python library that simplifies working with video from soccer matches. It allows match video to be selected intuitively by period number and clock, instead of absolute video time.

To accomplish this the start of each period is set as a chapter in the match video's metadata. Clips from the video can then be selected by period number and clock. ffmpeg handles both reading and writing the video chapter metadata and clip selection.

## Installation

### Requirements
- Python 3.6 or newer
- [ffmpeg](https://ffmpeg.org)

```shell
pip install match-video
```

## Usage

Before the video can be used, the start time of each half needs to be set.

```shell
match-video set-half-starts path/to/video.mp4 0:04 63:20
```

Then it is easy to select match video by period and clock!

```python
import match_video as mv

# get the third minute of the match
clip = mv.get_clip("path/to/video.mp4", period=1, start_clock=180, end_clock=240)

# get the start of each half and concatenate them
clip_clocks = [
    {"period": 1, "start_clock": 0, "end_clock": 30},
    {"period": 2, "start_clock": 0, "end_clock": 30},
]
clips = mv.get_clips("path/to/video.mp4", clip_clocks)
```

See the [examples](https://gitlab.com/grantwenzinger/match-video/-/tree/main/examples) to see how to save or display video clips.

## Support

<grantwenzinger@gmail.com>

## License

[MIT](https://choosealicense.com/licenses/mit/)


CHANGELOG.md
# Changelog
All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).

## [Unreleased]

## [0.1.0] - 2021-11-24
### Added
- Methods for reading and writing anchors to video files.
- Methods for getting clips from a match video with set anchors.
- Command-line interface for reading and writing anchors.
- Unit tests.
- Examples showing how to get and use clips in scripts, Jupyter notebooks, and Streamlit apps.


tests/test_cli.py
from unittest.mock import call, patch

import match_video.cli as cli
from match_video.anchor import Anchor


@patch("match_video.cli.utils.write_anchors")
def test_set_half_starts(mock_write_anchors):
    cli.set_half_starts("input_path", "0:00", "60:00", "output_path")

    mock_write_anchors.assert_called_once_with(
        "input_path",
        "output_path",
        [
            Anchor(1, 0.0, 0.0),
            Anchor(2, 0.0, 3600.0),
        ],
    )


@patch("match_video.cli.utils.write_anchors")
def test_set_half_starts_inplace(mock_write_anchors):
    cli.set_half_starts("path", "0:00", "60:00")

    mock_write_anchors.assert_called_once_with(
        "path",
        "path",
        [
            Anchor(1, 0.0, 0.0),
            Anchor(2, 0.0, 3600.0),
        ],
    )


@patch("match_video.cli.typer.echo")
@patch(
    "match_video.cli.utils.read_anchors",
    return_value=[
        Anchor(1, 0.0, 0.0),
        Anchor(2, 0.0, 3600.0),
    ],
)
def test_read_anchors(mock_read_anchors, mock_typer_echo):
    cli.read_anchors("path")

    expected_calls = [
        call("Period 1 0:00 | 0:00 in video"),
        call("Period 2 0:00 | 60:00 in video"),
    ]
    mock_typer_echo.assert_has_calls(expected_calls, any_order=False)


@patch("match_video.cli.typer.echo")
@patch("match_video.cli.utils.read_anchors", return_value=[])
def test_read_anchors_no_anchors(mock_read_anchors, mock_typer_echo):
    cli.read_anchors("path")

    mock_typer_echo.assert_called_once_with("No anchors set for video")


tests/test_utils.py
import json
import subprocess
from unittest.mock import MagicMock, patch

import pytest

import match_video.utils as utils
from match_video.anchor import Anchor


@patch("match_video.utils.NamedTemporaryFile")
@patch("os.path.exists", return_value=False)
@patch("subprocess.run")
def test_write_anchors_no_periods(
    mock_subprocess_run, mock_exists, mock_temp_file_context
):
    existing_metadata = """;FFMETADATA1
major_brand=brand
minor_version=1
compatible_brands=compatible
encoder=Lavf58.76.100"""

    mock_temp_file_context.return_value.__enter__.return_value.read.return_value = (
        existing_metadata
    )

    anchors = []

    utils.write_anchors("input_path", "output_path", anchors)

    mock_temp_file = mock_temp_file_context.return_value.__enter__.return_value
    mock_temp_file.write.assert_called_once_with(existing_metadata)


@patch("match_video.utils.NamedTemporaryFile")
@patch("os.path.exists", return_value=False)
@patch("subprocess.run")
def test_write_anchors_two_periods(
    mock_subprocess_run, mock_exists, mock_temp_file_context
):
    existing_metadata = """;FFMETADATA1
major_brand=brand
minor_version=1
compatible_brands=compatible
encoder=Lavf58.76.100"""

    mock_temp_file_context.return_value.__enter__.return_value.read.return_value = (
        existing_metadata
    )

    anchors = [
        Anchor(1, 0.0, 10.0),
        Anchor(2, 0.0, 1000.0),
    ]

    utils.write_anchors("input_path", "output_path", anchors)

    expected_metadata = """;FFMETADATA1
major_brand=brand
minor_version=1
compatible_brands=compatible
encoder=Lavf58.76.100

[CHAPTER]
TIMEBASE=1/1000
START=10000
END=10001
title=Period 1, 0.0

[CHAPTER]
TIMEBASE=1/1000
START=1000000
END=1000001
title=Period 2, 0.0"""

    mock_temp_file = mock_temp_file_context.return_value.__enter__.return_value
    mock_temp_file.write.assert_called_once_with(expected_metadata)


@patch("match_video.utils.NamedTemporaryFile")
@patch("os.path.exists", return_value=False)
@patch("subprocess.run")
def test_write_anchors_overwrite(
    mock_subprocess_run, mock_exists, mock_temp_file_context
):
    existing_metadata = """;FFMETADATA1
major_brand=brand
minor_version=1
compatible_brands=compatible
encoder=Lavf58.76.100

[CHAPTER]
TIMEBASE=1/1000
START=10000
END=10001
title=Period 1, 0.0

[CHAPTER]
TIMEBASE=1/1000
START=1000000
END=1000001
title=Period 2, 0.0"""

    mock_temp_file_context.return_value.__enter__.return_value.read.return_value = (
        existing_metadata
    )

    anchors = [
        Anchor(3, 10.0, 30.0),
    ]

    utils.write_anchors("input_path", "output_path", anchors)

    expected_metadata = """;FFMETADATA1
major_brand=brand
minor_version=1
compatible_brands=compatible
encoder=Lavf58.76.100

[CHAPTER]
TIMEBASE=1/1000
START=30000
END=30001
title=Period 3, 10.0"""

    mock_temp_file = mock_temp_file_context.return_value.__enter__.return_value
    mock_temp_file.write.assert_called_once_with(expected_metadata)


@patch("shutil.copy2")
@patch("match_video.utils.NamedTemporaryFile")
@patch("os.path.samefile", return_value=True)
@patch("os.path.exists", return_value=True)
@patch("subprocess.run")
def test_write_anchors_inplace(
    mock_subprocess_run, mock_exists, mock_samefile, mock_temp_file_context, mock_copy2
):
    existing_metadata = """;FFMETADATA1
major_brand=brand
minor_version=1
compatible_brands=compatible
encoder=Lavf58.76.100"""

    mock_temp_file_context.return_value.__enter__.return_value.read.return_value = (
        existing_metadata
    )
    mock_temp_file_context.return_value.__enter__.return_value.name = (
        "intermediate_file_name"
    )

    anchors = [
        Anchor(1, 0.0, 10.0),
        Anchor(2, 0.0, 1000.0),
    ]

    utils.write_anchors("path", "path", anchors)

    expected_metadata = """;FFMETADATA1
major_brand=brand
minor_version=1
compatible_brands=compatible
encoder=Lavf58.76.100

[CHAPTER]
TIMEBASE=1/1000
START=10000
END=10001
title=Period 1, 0.0

[CHAPTER]
TIMEBASE=1/1000
START=1000000
END=1000001
title=Period 2, 0.0"""

    mock_temp_file = mock_temp_file_context.return_value.__enter__.return_value
    mock_temp_file.write.assert_called_once_with(expected_metadata)

    mock_copy2.assert_called_once_with("intermediate_file_name", "path")


@patch("subprocess.run")
def test_read_anchors_no_periods(mock_subprocess_run):
    mock_subprocess_run.return_value = MagicMock()
    mock_subprocess_run.return_value.stdout = json.dumps({"chapters": []})

    result = utils.read_anchors("path")

    mock_subprocess_run.assert_called_once_with(
        [
            "ffprobe",
            "-v",
            "quiet",
            "-print_format",
            "json",
            "-show_error",
            "-show_chapters",
            "path",
        ],
        stdout=subprocess.PIPE,
        stderr=subprocess.PIPE,
    )
    assert len(result) == 0


@patch("subprocess.run")
def test_read_anchors_one_period(mock_subprocess_run):
    mock_subprocess_run.return_value = MagicMock()
    mock_subprocess_run.return_value.stdout = json.dumps(
        {
            "chapters": [
                {
                    "id": 0,
                    "time_base": "1/1000",
                    "start": 0,
                    "start_time": "0.000000",
                    "end": 1000000,
                    "end_time": "1000.000000",
                    "tags": {"title": "Period 1, 0.0"},
                }
            ]
        }
    )

    result = utils.read_anchors("path")

    mock_subprocess_run.assert_called_once_with(
        [
            "ffprobe",
            "-v",
            "quiet",
            "-print_format",
            "json",
            "-show_error",
            "-show_chapters",
            "path",
        ],
        stdout=subprocess.PIPE,
        stderr=subprocess.PIPE,
    )
    assert len(result) == 1

    assert result[0] == Anchor(1, 0.0, 0.0)


@patch("subprocess.run")
def test_read_anchors_many_periods(mock_subprocess_run):
    mock_subprocess_run.return_value = MagicMock()
    mock_subprocess_run.return_value.stdout = json.dumps(
        {
            "chapters": [
                {
                    "id": 0,
                    "time_base": "1/1000",
                    "start": 0,
                    "start_time": "0.000000",
                    "end": 1000000,
                    "end_time": "1000.000000",
                    "tags": {"title": "Period 1, 0.0"},
                },
                {
                    "id": 0,
                    "time_base": "1/1000",
                    "start": 1100000,
                    "start_time": "1100.000000",
                    "end": 2100000,
                    "end_time": "2100.000000",
                    "tags": {"title": "Period 2, 10.0"},
                },
                {
                    "id": 0,
                    "time_base": "1/1000",
                    "start": 5000000,
                    "start_time": "5000.000000",
                    "end": 5100000,
                    "end_time": "5100.000000",
                    "tags": {"title": "Period 3, 0.0"},
                },
                {
                    "id": 0,
                    "time_base": "1/1000",
                    "start": 5200000,
                    "start_time": "5200.000000",
                    "end": 5300000,
                    "end_time": "5300.000000",
                    "tags": {"title": "Period 4, 0.0"},
                },
            ]
        }
    )

    result = utils.read_anchors("path")

    mock_subprocess_run.assert_called_once_with(
        [
            "ffprobe",
            "-v",
            "quiet",
            "-print_format",
            "json",
            "-show_error",
            "-show_chapters",
            "path",
        ],
        stdout=subprocess.PIPE,
        stderr=subprocess.PIPE,
    )
    assert len(result) == 4

    assert result[0] == Anchor(1, 0.0, 0.0)
    assert result[1] == Anchor(2, 10.0, 1100.0)
    assert result[2] == Anchor(3, 0.0, 5000.0)
    assert result[3] == Anchor(4, 0.0, 5200.0)


@patch("subprocess.run")
def test_read_anchors_bad_path(mock_subprocess_run):
    mock_subprocess_run.return_value = MagicMock()
    mock_subprocess_run.return_value.stdout = json.dumps(
        {"error": {"code": -2, "string": "No such file or directory"}}
    )

    with pytest.raises(ValueError):
        utils.read_anchors("path")


@patch("match_video.utils._extract_clip")
@patch("match_video.utils.NamedTemporaryFile")
@patch("match_video.utils._video_sans_chapters")
@patch(
    "match_video.utils.read_anchors",
    return_value=[
        Anchor(1, 0.0, 0.0),
        Anchor(2, 0.0, 1000.0),
    ],
)
def test_get_clip(
    mock_read_anchors,
    mock_video_sans_chapters_context,
    mock_temp_file_context,
    mock_extract_clip,
):
    clip = utils.get_clip("path", 1, 0.0, 10.0)

    video_sans_chapters_path = (
        mock_video_sans_chapters_context.return_value.__enter__.return_value
    )
    clip_file_path = mock_temp_file_context.return_value.__enter__.return_value.name

    mock_extract_clip.assert_called_once_with(
        video_sans_chapters_path, clip_file_path, 0.0, 10.0
    )


@patch("match_video.utils.read_anchors", return_value=[])
def test_get_clip_no_anchors(mock_read_anchors):
    with pytest.raises(ValueError):
        utils.get_clip("path", 1, 0.0, 10.0)


@patch("match_video.utils.read_anchors", return_value=[])
def test_get_clips_no_anchors(mock_read_anchors):
    clip_clocks = [
        {"period": 1, "start_clock": 0.0, "end_clock": 10.0},
    ]

    with pytest.raises(ValueError):
        utils.get_clips("path", clip_clocks)


def test_get_video_time_half_start():
    anchors = [
        Anchor(1, 0.0, 100.0),
        Anchor(2, 0.0, 10000.0),
    ]

    start_video_time, end_video_time = utils._get_video_times(anchors, 1, 0.0, 10.0)

    assert start_video_time == 100.0
    assert end_video_time == 110.0


def test_get_video_time_offset():
    anchors = [
        Anchor(1, 0.0, 100.0),
        Anchor(2, 0.0, 10000.0),
    ]

    start_video_time, end_video_time = utils._get_video_times(anchors, 1, 670.0, 1000.0)

    assert start_video_time == 770.0
    assert end_video_time == 1100.0


def test_get_video_time_later_period():
    anchors = [
        Anchor(1, 0.0, 100.0),
        Anchor(2, 0.0, 10000.0),
        Anchor(3, 0.0, 20000.0),
        Anchor(4, 0.0, 30000.0),
    ]

    start_video_time, end_video_time = utils._get_video_times(anchors, 4, 100, 1000.0)

    assert start_video_time == 30100.0
    assert end_video_time == 31000.0


def test_get_video_time_before_discontinuity():
    anchors = [
        Anchor(1, 0.0, 100.0),
        Anchor(1, 1000.0, 2000.0),
        Anchor(2, 0.0, 10000.0),
    ]

    start_video_time, end_video_time = utils._get_video_times(anchors, 1, 500.0, 600.0)

    assert start_video_time == 600.0
    assert end_video_time == 700.0


def test_get_video_time_after_discontinuity():
    anchors = [
        Anchor(1, 0.0, 100.0),
        Anchor(1, 1000.0, 2000.0),
        Anchor(2, 0.0, 10000.0),
    ]

    start_video_time, end_video_time = utils._get_video_times(
        anchors, 1, 1500.0, 1600.0
    )

    assert start_video_time == 2500.0
    assert end_video_time == 2600.0


def test_get_video_time_before_first_anchor_in_period():
    anchors = [
        Anchor(1, 1000.0, 2000.0),
        Anchor(2, 0.0, 10000.0),
    ]

    with pytest.raises(ValueError):
        utils._get_video_times(anchors, 1, 10.0, 20.0)


match_video/cli.py
from typing import Optional

import typer

import match_video.utils as utils
from match_video.anchor import Anchor

app = typer.Typer()


@app.command()
def set_half_starts(
    input_video_path: str,
    first_half_start_time: str,
    second_half_start_time: str,
    output_video_path: Optional[str] = None,
) -> None:
    """Set the start times for each half of a match video.

    Args:
        input_video_path: The path to a video.
        first_half_start_time: The start of the first half as mm:ss in video time.
        second_half_start_time: The start of the second half as mm:ss in video time.
        output_video_path: The path to write the video with anchors to. Overwrite the
            input video if this is not specified.
    """
    if output_video_path is None:
        output_video_path = input_video_path

    def get_anchor(period: int, video_time_str: str) -> Anchor:
        video_minutes, video_seconds = video_time_str.split(":")
        video_time = int(video_minutes) * 60 + int(video_seconds)

        return Anchor(period, 0.0, video_time)

    anchors = [
        get_anchor(1, first_half_start_time),
        get_anchor(2, second_half_start_time),
    ]

    utils.write_anchors(input_video_path, output_video_path, anchors)


@app.command()
def read_anchors(video_path: str) -> None:
    """Read the anchors set for a video.

    Args:
        video_path: The path to a video.
    """
    anchors = utils.read_anchors(video_path)

    if len(anchors) == 0:
        typer.echo("No anchors set for video")

    for anchor in anchors:
        minute = int(anchor.clock / 60)
        second = int(anchor.clock % 60)

        video_minute = int(anchor.video_time / 60)
        video_second = int(anchor.video_time % 60)

        typer.echo(
            f"Period {anchor.period} {minute}:{second:02} | {video_minute}:{video_second:02} in video"
        )


if __name__ == "__main__":
    app()


match_video/utils.py
import json
import os
import shutil
import subprocess
from contextlib import contextmanager
from operator import attrgetter
from tempfile import NamedTemporaryFile
from typing import Iterator, List, Tuple

from match_video.anchor import Anchor


def write_anchors(
    input_video_path: str, output_video_path: str, anchors: List[Anchor]
) -> None:
    """Write anchors to a video file.

    Anchors are written as chapters in the video's metadata. Any existing anchors will
    be overwritten.

    Args:
        input_video_path: The path to a video.
        output_video_path: The path to write the video with anchors to.
        anchors: A list of anchors specifying the start of each half and any
            discontinuities in the video.
    """
    existing_metadata: str

    with NamedTemporaryFile("r") as existing_metadata_file:
        subprocess.run(
            [
                "ffmpeg",
                "-y",
                "-i",
                input_video_path,
                "-f",
                "ffmetadata",
                existing_metadata_file.name,
            ],
            stdout=subprocess.PIPE,
            stderr=subprocess.PIPE,
        )

        existing_metadata_file.seek(0)
        existing_metadata = existing_metadata_file.read()

    existing_metadata_without_chapters = (
        existing_metadata.split("[CHAPTER]")[0].strip()
        if "[CHAPTER]" in existing_metadata
        else existing_metadata
    )

    updated_metadata = existing_metadata_without_chapters

    for anchor in anchors:
        time = int(anchor.video_time * 1000)

        chapter = f"""

[CHAPTER]
TIMEBASE=1/1000
START={time}
END={time + 1}
title=Period {anchor.period}, {anchor.clock}"""

        updated_metadata += chapter

    with NamedTemporaryFile("w") as updated_metadata_file:
        updated_metadata_file.write(updated_metadata)
        updated_metadata_file.seek(0)

        def write_video_with_metadata(path: str) -> None:
            subprocess.run(
                [
                    "ffmpeg",
                    "-y",
                    "-i",
                    input_video_path,
                    "-i",
                    updated_metadata_file.name,
                    "-map_metadata",
                    "1",
                    "-codec",
                    "copy",
                    path,
                ],
                stdout=subprocess.PIPE,
                stderr=subprocess.PIPE,
            )

        if os.path.exists(output_video_path) and os.path.samefile(
            input_video_path, output_video_path
        ):
            # ffmpeg can't update video metadata in place, so use an intermediate file

            with NamedTemporaryFile("w", suffix=".mp4") as intermediate_file:
                write_video_with_metadata(intermediate_file.name)
                shutil.copy2(intermediate_file.name, output_video_path)
        else:
            write_video_with_metadata(output_video_path)


def read_anchors(video_path: str) -> List[Anchor]:
    """Read the anchor points from the chapter metadata of a video file.

    Args:
        video_path: The path to a video.

    Returns:
        The list of anchors set for the video.

    Raises:
        ValueError: The video's metadata could not be read.
    """
    result_json = subprocess.run(
        [
            "ffprobe",
            "-v",
            "quiet",
            "-print_format",
            "json",
            "-show_error",
            "-show_chapters",
            video_path,
        ],
        stdout=subprocess.PIPE,
        stderr=subprocess.PIPE,
    )
    result = json.loads(result_json.stdout)

    if "error" in result:
        raise ValueError(
            f"Unable to read the metadata for {video_path}, {result['error']['string']}"
        )

    def get_anchor(chapter: dict) -> Anchor:
        period_str, clock_str = chapter["tags"]["title"].split(", ")
        period = int(period_str.split("Period ")[-1])
        clock = float(clock_str)

        video_time = float(chapter["start_time"])

        return Anchor(period, clock, video_time)

    anchors = [get_anchor(chapter) for chapter in result["chapters"]]

    return anchors


def get_clip(
    video_path: str, period: int, start_clock: float, end_clock: float
) -> bytes:
    """Get a clip from a match by period and clock.

    Args:
        video_path: The path to a video.
        period: The period of the match the clip is in.
        start_clock: The start of the clip in seconds since the start of the period.
        end_clock: The end of the clip in seconds since the start of the period.

    Returns:
        The video clip as bytes.

    Raises:
        ValueError: The video does not have anchors or the clips is before the first
            anchor in its period.
    """
    anchors = read_anchors(video_path)

    if len(anchors) == 0:
        raise ValueError(f"{video_path} has no set anchors")

    start_video_time, end_video_time = _get_video_times(
        anchors, period, start_clock, end_clock
    )

    clip: bytes

    with _video_sans_chapters(video_path) as video_sans_chapters_path:
        with NamedTemporaryFile("rb", suffix=".mp4") as clip_file:
            _extract_clip(
                video_sans_chapters_path,
                clip_file.name,
                start_video_time,
                end_video_time,
            )

            clip = clip_file.read()

    return clip


def get_clips(video_path: str, clip_clocks: List[dict]) -> bytes:
    """Get clips from a match by period and clock.

    Args:
        video_path: The path to a video.
        clip_clocks: A list of clips to select and stitch together. Each clip
            dictionary should have a period, start_clock, and end_clock. These values
            are the same as with get_clip.

    Returns:
        The video clips as bytes.

    Raises:
        ValueError: The video does not have anchors or one of the clips is before the
            first anchor in its period.
    """
    anchors = read_anchors(video_path)

    if len(anchors) == 0:
        raise ValueError(f"{video_path} has no set anchors")

    clip_files: List

    with _video_sans_chapters(video_path) as video_sans_chapters_path:

        def create_clip_file(clip_info: dict):
            clip_file = NamedTemporaryFile("rb", suffix=".mp4")

            start_video_time, end_video_time = _get_video_times(
                anchors,
                clip_info["period"],
                clip_info["start_clock"],
                clip_info["end_clock"],
            )

            _extract_clip(
                video_sans_chapters_path,
                clip_file.name,
                start_video_time,
                end_video_time,
            )

            return clip_file

        clip_files = [create_clip_file(clip_info) for clip_info in clip_clocks]

    clips: bytes

    with NamedTemporaryFile("w") as clip_paths_file:
        clip_paths = "\n".join([f"file '{clip_file.name}'" for clip_file in clip_files])
        clip_paths_file.write(clip_paths)
        clip_paths_file.seek(0)

        with NamedTemporaryFile("rb", suffix=".mp4") as clips_file:
            subprocess.run(
                [
                    "ffmpeg",
                    "-y",
                    "-f",
                    "concat",
                    "-safe",
                    "0",
                    "-i",
                    clip_paths_file.name,
                    "-c",
                    "copy",
                    clips_file.name,
                ],
                stdout=subprocess.PIPE,
                stderr=subprocess.PIPE,
            )

            clips = clips_file.read()

    for clip_file in clip_files:
        clip_file.close()

    return clips


def _get_video_times(
    anchors: List[Anchor], period: int, start_clock: float, end_clock: float
) -> Tuple[float, float]:
    """Convert period and match clocks into video times.

    Args:
        anchors: A video's anchors.
        period: The period of the match.
        start_clock: A time since the start of the period.
        end_clock: A time since the start of the period.

    Returns:
        A pair, (start_video_time, end_video_time), that correspond to start_clock and
        end_clock in video time.

    Raises:
        ValueError: There are no anchors in the period before start_clock.
    """
    prior_anchors = sorted(
        [
            anchor
            for anchor in anchors
            if anchor.period == period and anchor.clock <= start_clock
        ],
        key=attrgetter("clock"),
    )

    if len(prior_anchors) == 0:
        minute = int(start_clock / 60)
        second = int(start_clock % 60)

        raise ValueError(
            f"No anchors set in period {period} before {minute}:{second:02}"
        )

    last_anchor = prior_anchors[-1]
    start_offset = start_clock - last_anchor.clock
    start_video_time = last_anchor.video_time + start_offset

    duration = end_clock - start_clock
    end_video_time = start_video_time + duration

    return start_video_time, end_video_time


@contextmanager
def _video_sans_chapters(video_path: str) -> Iterator[str]:
    """Create a copy of a video without its chapters.

    Args:
        video_path: The path to a video.

    Yields:
        The path to the copied video file without chapters.
    """
    with NamedTemporaryFile("rb", suffix=".mp4") as video_sans_chapters_file:
        subprocess.run(
            [
                "ffmpeg",
                "-y",
                "-i",
                video_path,
                "-map",
                "0",
                "-vcodec",
                "copy",
                "-acodec",
                "copy",
                "-map_chapters",
                "-1",
                video_sans_chapters_file.name,
            ],
            stdout=subprocess.PIPE,
            stderr=subprocess.PIPE,
        )

        yield video_sans_chapters_file.name


def _extract_clip(
    input_video_path: str, output_video_path: str, start_time: float, end_time: float
) -> None:
    """Extract a clip from input_video_path and write it to output_video_path.

    Output can be incorrect if input_video_path has chapters. Use _video_sans_chapters
    to copy the video file without chapters before using this method.

    Args:
        input_video_path: The path to a video.
        output_video_path: The path to write the video with anchors to.
        start_time: The start of the clip in seconds since video start.
        end_time: The end of the clip in seconds since video start.
    """
    subprocess.run(
        [
            "ffmpeg",
            "-y",
            "-ss",
            f"{start_time:0.2f}",
            "-to",
            f"{end_time:0.2f}",
            "-i",
            input_video_path,
            "-map",
            "0",
            "-vcodec",
            "copy",
            "-acodec",
            "copy",
            output_video_path,
        ],
        stdout=subprocess.PIPE,
        stderr=subprocess.PIPE,
    )
